tsdb/head_test.go

```diff
@@ -247,21 +247,87 @@ func BenchmarkLoadWAL(b *testing.B) {
 					b.ResetTimer()
 
 					// Load the WAL.
+					ctx, cancel := context.WithCancel(context.Background())
 					for i := 0; i < b.N; i++ {
 						opts := DefaultHeadOptions()
 						opts.ChunkRange = 1000
 						opts.ChunkDirRoot = w.Dir()
 						h, err := NewHead(nil, nil, w, opts, nil)
 						require.NoError(b, err)
-						h.Init(0)
+						h.Init(ctx, 0)
 					}
 					b.StopTimer()
 					w.Close()
+					cancel()
 				})
 		}
 	}
 }
 
+func TestLoadWALCancel(t *testing.T) {
+	labelsPerSeries := 5
+	batches := 100
+	seriesPerBatch := 10000
+	samplesPerSeries := 100
+
+	dir := t.TempDir()
+
+	w, err := wal.New(nil, nil, dir, false)
+	require.NoError(t, err)
+
+	// Write series.
+	refSeries := make([]record.RefSeries, 0, seriesPerBatch)
+	for k := 0; k < batches; k++ {
+		refSeries = refSeries[:0]
+		for i := k * seriesPerBatch; i < (k+1)*seriesPerBatch; i++ {
+			lbls := make(map[string]string, labelsPerSeries)
+			lbls[defaultLabelName] = strconv.Itoa(i)
+			for j := 1; len(lbls) < labelsPerSeries; j++ {
+				lbls[defaultLabelName+strconv.Itoa(j)] = defaultLabelValue + strconv.Itoa(j)
+			}
+			refSeries = append(refSeries, record.RefSeries{Ref: chunks.HeadSeriesRef(i) * 101, Labels: labels.FromMap(lbls)})
+		}
+		populateTestWAL(t, w, []interface{}{refSeries})
+	}
+
+	// Write samples.
+	refSamples := make([]record.RefSample, 0, seriesPerBatch)
+	for i := 0; i < samplesPerSeries; i++ {
+		for j := 0; j < batches; j++ {
+			refSamples = refSamples[:0]
+			for k := j * seriesPerBatch; k < (j+1)*seriesPerBatch; k++ {
+				refSamples = append(refSamples, record.RefSample{
+					Ref: chunks.HeadSeriesRef(k) * 101,
+					T:   int64(i) * 10,
+					V:   float64(i) * 100,
+				})
+			}
+			populateTestWAL(t, w, []interface{}{refSamples})
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	opts := DefaultHeadOptions()
+	opts.ChunkRange = 1000
+	opts.ChunkDirRoot = w.Dir()
+	h, err := NewHead(nil, nil, w, opts, nil)
+	require.NoError(t, err)
+
+	// Call cancel after 50 milliseconds while WAL is being loaded.
+	go func() {
+		time.Sleep(time.Millisecond * 50)
```

Comment:
Not entirely sure how to test it better, this works but it depends on timings. Hopefully 50ms is safe here.
The trick is that we need to `cancel()`  while WAL is being loaded, but I don't know how to be sure that we've started loading it other than just wait a little.